### DISCUSSION TITLE
How can I use "import A from B" style python import in C#?

### DISCUSSION THREAD

**Discussion by hrntsm** [pythonnet/pythonnet#2138] [Q&A]
Hi discussions!

How can I use "import A from B" style python import in C#?

```py
from math import radians
```

Is the following the only way to write?

```cs
using (Py.GIL())
{
    PyModule ps = Py.CreateScope();
    ps.Exec(
        "from math import radians"
    );
    dynamic radians = ps.Get("radians");
}
```

Is there a function available like  `Py.Import("radians", "math")`?

Thanks in advance.

**Comment by filmor** [comment]
`Py.Import("math").Get("radians")` should work.

**Reply by hrntsm** [reply]
It worked on `Py.Import("math").GetAttr("radians")`  (not `Get`)
Thank you.

**Comment by cigoic** [comment]
hi,
I have the similar question: `from A.B.C import D`
How do I write this in .NET Core 6 ?

I tried coding it in C#: 

`dynamic D = Py.Import("A").GetAttr("B").GetAttr("C").GetAttr("D");`
or
`dynamic D = Py.Import("A.B.C").GetAttr("D");`

None of them work.
My app pops up a wrong message shown that it cannot find the module A.
I won't be able to get correct configuration under macOS.

```csharp
var pathToVirtualEnv = @"/Users/jmosx/__path_to_my_python_project__/.venv";
string pythonVersion = @"python3.11";
Runtime.PythonDLL = $@"/Users/jmosx/miniconda3/envs/__virtual_environment_name__/lib/lib{pythonVersion}.dylib";
PythonEngine.Initialize();
```

It did execute the python.net sample code using above configuration but not work for me if I try to import D from A.B.C.

Therefore, I followed the instruction from https://github.com/pythonnet/pythonnet/wiki/Using-Python.NET-with-Virtual-Environments
Setting other paths but fail again as show below. The .NET Core 6 MVC crashed itself executing PythonEngine.Initialize() every time without specific error message.

```csharp
      string pythonHome = @"/Users/jmosx/path_to_python_project/.venv";
      string pythonVersion = @"python3.11";
      string pythonCondaPath = @"/Users/jmosx/miniconda3/envs/virtual_enironment_name";
      Runtime.PythonDLL = $@"{pythonCondaPath}/lib/lib{pythonVersion}.dylib";
      string pythonPath = $@"{pythonHome}/lib/{pythonVersion}/site-packages:{pythonCondaPath}/lib/{pythonVersion}";


      PythonEngine.PythonHome = pythonHome;
      PythonEngine.PythonPath = pythonPath;

      try
      {
        PythonEngine.Initialize();
        _logger.LogInformation("Python Version: {v}, {dll}", PythonEngine.Version.Trim(), Runtime.PythonDLL);
        _logger.LogInformation("Python Home: {home}", PythonEngine.PythonHome);
        _logger.LogInformation("Python Path: {path}", PythonEngine.PythonPath);
      }
      catch (PythonException e)
      {
        throw new Exception($"Python Version: {PythonEngine.Version.Trim()}, {Runtime.PythonDLL}, Python Exception, {e.Message}", e);
      }
      catch (System.TypeInitializationException e)
      {
        throw new Exception($"FATAL, Unable to load Python, dll={Runtime.PythonDLL}", e);
      }
      catch (Exception e)
      {
        throw new Exception($"Python initialization Exception, {e.Message}", e);
      }
```

Any suggestion?

**Reply by hrntsm** [reply]
@cigoic 
What python code do you actually expect?
The structure of the files is slightly different depending on the python library, so if you can tell me the python code and the library that reads the actual library, I might be able to tell you something.

**Reply by cigoic** [reply]
hi, @hrntsm 

It's the [Iangchain](https://python.langchain.com/) package. For example,

```py
from langchain.document_loaders import TextLoader
from langchain.embeddings.openai import OpenAIEmbeddings
```

I saw another tutorial online showing the usage of PythonInterop.RunPythonCode(). What's the difference?
Would you recommend using this instead of Py.GIL()?

Thanks!
